CSS TypedOM crash under BuilderFunctions::applyValueTranslate

anttijk opened the pull request:
#### 00b74e1a7b8cb148b8500e731132c1d881c21ab4
<pre>
CSS TypedOM crash under BuilderFunctions::applyValueTranslate
<a href="https://bugs.webkit.org/show_bug.cgi?id=294670">https://bugs.webkit.org/show_bug.cgi?id=294670</a>
<a href="https://rdar.apple.com/146650025">rdar://146650025</a>

Reviewed by Darin Adler.

Use BuilderConverter::requiredDowncast&lt;&gt; to check for correct types when building transforms.

* LayoutTests/fast/css/cssom-translate-crash-expected.txt: Added.
* LayoutTests/fast/css/cssom-translate-crash.html: Added.
* Source/WebCore/css/DOMMatrixReadOnly.cpp:
(WebCore::DOMMatrixReadOnly::parseStringIntoAbstractMatrix):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeTypedCustomPropertyValue):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.cpp:
(WebCore::CSSPropertyParserHelpers::parseTransformRaw):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.h:
* Source/WebCore/css/query/ContainerQueryFeatures.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTransform):
* Source/WebCore/style/StyleBuilderState.h:
(WebCore::Style::BuilderState::parentStyle const):
(WebCore::Style::BuilderState::document const):
(WebCore::Style::BuilderState::protectedDocument const):

Make it simpler to construct a dummy BuilderState without Builder.

* Source/WebCore/style/StyleCustomPropertyRegistry.cpp:
(WebCore::Style::CustomPropertyRegistry::parseInitialValue):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::builderContext):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveAgainInDifferentContext):
(WebCore::Style::TreeResolver::applyCascadeAfterAnimation):
* Source/WebCore/style/TransformOperationsBuilder.cpp:
(WebCore::Style::createMatrixTransformOperation):
(WebCore::Style::createMatrix3dTransformOperation):
(WebCore::Style::createRotateTransformOperation):
(WebCore::Style::createRotate3dTransformOperation):
(WebCore::Style::createRotateXTransformOperation):
(WebCore::Style::createRotateYTransformOperation):
(WebCore::Style::createRotateZTransformOperation):
(WebCore::Style::createSkewTransformOperation):
(WebCore::Style::createSkewXTransformOperation):
(WebCore::Style::createSkewYTransformOperation):
(WebCore::Style::createScaleTransformOperation):
(WebCore::Style::createScale3dTransformOperation):
(WebCore::Style::createScaleXTransformOperation):
(WebCore::Style::createScaleYTransformOperation):
(WebCore::Style::createScaleZTransformOperation):
(WebCore::Style::createTranslateTransformOperation):
(WebCore::Style::createTranslate3dTransformOperation):
(WebCore::Style::createTranslateXTransformOperation):
(WebCore::Style::createTranslateYTransformOperation):
(WebCore::Style::createTranslateZTransformOperation):
(WebCore::Style::createPerspectiveTransformOperation):

Pass BuilderState to various create* functions and use requiredFunctionDowncast&lt;&gt; for checked safe downcasts.
Return a RefPtr instead of a Ref.

(WebCore::Style::createTransformOperation):
(WebCore::Style::createTransformOperations):
(WebCore::Style::resolveAsNumberAtIndex): Deleted.
* Source/WebCore/style/TransformOperationsBuilder.h:

Canonical link: <a href="https://commits.webkit.org/296528@main">https://commits.webkit.org/296528@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da024d5b708346a46247471c414dbea460a2d84a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108818 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28478 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18903 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114026 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/59179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110781 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29162 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37042 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/82679 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/59179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111766 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23170 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98010 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63118 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22593 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/16148 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58730 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/92541 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16196 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117147 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35868 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26491 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91699 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36241 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94277 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91509 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23294 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36402 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14158 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31742 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35768 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/41300 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35470 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38815 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37155 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.cpp:
```diff
@@ -429,10 +431,13 @@ std::optional<TransformOperations> parseTransformRaw(const String& string, const
         return { };
 
-    if (!parsedValue->canResolveDependenciesWithConversionData(conversionData))
+    auto dummyStyle = RenderStyle::create();
+    auto dummyState = Style::BuilderState { dummyStyle };
+
+    if (!parsedValue->canResolveDependenciesWithConversionData(dummyState.cssToLengthConversionData()))
```
This seems a bit inelegant. Will we always want to do this or is this a temporary stopgap that we can eventually avoid?

weinig replied:
`parseTransformRaw` is a bit of an odd duck and exists for only one caller, `DOMMatrixReadOnly::parseStringIntoAbstractMatrix`. `DOMMatrixReadOnly::parseStringIntoAbstractMatrix` has the following in its spec (https://drafts.fxtf.org/geometry/#dommatrix-parse):

> 2. Parse transformList into parsedValue given the grammar for the CSS transform property. The result will be a `<transform-list>`, the keyword none, or failure. If parsedValue is failure, or any `<transform-function>` has `<length>` values without absolute length units, or any keyword other than none is used, then return failure. [CSS3-SYNTAX] [CSS3-TRANSFORMS]

So what we are trying to do here, in a round-about way, is check that no font-relative, viewport-relative, or container-relative units are used. 

I think this, parse like normal, and then check if the result had disallowed units, is the wrong way to go about it.

Instead, I would propose a new regime.

I think we should make it so that length resolution will **always succeed**. That is, `CSSToLengthConversionData` always has valid data for converting a `rex` or `dvw` to pixels, even in the cases where there is no RenderStyle, root RenderStyle, element, document or RenderView. 

- `font-relative` and `root font-relative` units will resolve against the "initial values of the font and line-height properties" if there are missing RenderStyles.
- `viewport-relative` units will use initial value of the `writing-mode` property if no element or style available, and the current [0 or value] they use if no RenderView is available.
- `container-relative` will fallback on the viewport (and its fallback) if there is no element container (as it does today).

Instead, for the cases where these things are explicitly not allowed, we will augment the CSS::ParserState with a new `only allow absolute lengths` (or maybe "computationally independent") bit that fails the parse when non-absolute lengths are used.

The benefit I see to this new regime is that:
- Makes things like this check much more explicit, it is now a bit on the parser state.
- Makes a whole class of crashes go away (see https://github.com/WebKit/WebKit/pull/46741).

--

All of this is a long winded way to say I think the general approach here is good, and we can go one step further after this by:

1. Updating computeNonCalcLengthDouble/computeUnzoomedNonCalcLengthDouble to implement fallback as noted above and never assert/crash.

2. Add bit to ParserState for cases that don't want to allow relative lengths, and then remove canResolveDependenciesWithConversionData() checks. (This might also have the benefit of removing a double parse in the custom properties code, though no promises).

darinadler commented on Source/WebCore/style/StyleBuilderState.h:
```diff
@@ -79,26 +79,27 @@ struct BuilderPositionTryFallback {
+    const RefPtr<const Document> document { };
+    const RenderStyle* parentStyle { nullptr };
+    const RenderStyle* rootElementStyle { nullptr };
+    RefPtr<const Element> element { nullptr };
```
Can these either be all `{ }` or all `{ nullptr }` or is there a reason to mix them?

darinadler commented on Source/WebCore/style/TransformOperationsBuilder.cpp:
```diff
@@ -108,238 +116,353 @@ static Ref<TransformOperation> createMatrix3dTransformOperation(const CSSFunctio
+    auto& conversionData = builderState.cssToLengthConversionData();
+
+
     auto tx = WebCore::Length(0, LengthType::Fixed);
```
```suggestion
    auto& conversionData = builderState.cssToLengthConversionData();

    auto tx = WebCore::Length(0, LengthType::Fixed);
```

anttijk committed:
CSS TypedOM crash under BuilderFunctions::applyValueTranslate
https://bugs.webkit.org/show_bug.cgi?id=294670
rdar://146650025

Reviewed by Darin Adler.

Use BuilderConverter::requiredDowncast<> to check for correct types when building transforms.

* LayoutTests/fast/css/cssom-translate-crash-expected.txt: Added.
* LayoutTests/fast/css/cssom-translate-crash.html: Added.
* Source/WebCore/css/DOMMatrixReadOnly.cpp:
(WebCore::DOMMatrixReadOnly::parseStringIntoAbstractMatrix):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeTypedCustomPropertyValue):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.cpp:
(WebCore::CSSPropertyParserHelpers::parseTransformRaw):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.h:
* Source/WebCore/css/query/ContainerQueryFeatures.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTransform):
* Source/WebCore/style/StyleBuilderState.h:
(WebCore::Style::BuilderState::parentStyle const):
(WebCore::Style::BuilderState::document const):
(WebCore::Style::BuilderState::protectedDocument const):

Make it simpler to construct a dummy BuilderState without Builder.

* Source/WebCore/style/StyleCustomPropertyRegistry.cpp:
(WebCore::Style::CustomPropertyRegistry::parseInitialValue):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::builderContext):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveAgainInDifferentContext):
(WebCore::Style::TreeResolver::applyCascadeAfterAnimation):
* Source/WebCore/style/TransformOperationsBuilder.cpp:
(WebCore::Style::createMatrixTransformOperation):
(WebCore::Style::createMatrix3dTransformOperation):
(WebCore::Style::createRotateTransformOperation):
(WebCore::Style::createRotate3dTransformOperation):
(WebCore::Style::createRotateXTransformOperation):
(WebCore::Style::createRotateYTransformOperation):
(WebCore::Style::createRotateZTransformOperation):
(WebCore::Style::createSkewTransformOperation):
(WebCore::Style::createSkewXTransformOperation):
(WebCore::Style::createSkewYTransformOperation):
(WebCore::Style::createScaleTransformOperation):
(WebCore::Style::createScale3dTransformOperation):
(WebCore::Style::createScaleXTransformOperation):
(WebCore::Style::createScaleYTransformOperation):
(WebCore::Style::createScaleZTransformOperation):
(WebCore::Style::createTranslateTransformOperation):
(WebCore::Style::createTranslate3dTransformOperation):
(WebCore::Style::createTranslateXTransformOperation):
(WebCore::Style::createTranslateYTransformOperation):
(WebCore::Style::createTranslateZTransformOperation):
(WebCore::Style::createPerspectiveTransformOperation):

Pass BuilderState to various create* functions and use requiredFunctionDowncast<> for checked safe downcasts.
Return a RefPtr instead of a Ref.

(WebCore::Style::createTransformOperation):
(WebCore::Style::createTransformOperations):
(WebCore::Style::resolveAsNumberAtIndex): Deleted.
* Source/WebCore/style/TransformOperationsBuilder.h:

Canonical link: https://commits.webkit.org/296528@main
